File: python/Q1584_Min_Cost_to_Connect_All_Points.py
class Solution:
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        from queue import PriorityQueue
        caldist = lambda p1, p2 : abs(p1[0]-p2[0])+abs(p1[1]-p2[1])

        pq = PriorityQueue()
        toVisit = set(range(len(points)))
        res = 0

        pq.put([0,0])

        while toVisit:
            cost, cur = pq.get()
            if cur not in toVisit:
                continue
            # not visited
            res += cost
            toVisit.remove(cur)

            for node in toVisit:
                pq.put([caldist(points[node],points[cur]), node])

        return res



    def minCostConnectPoints2(self, points: List[List[int]]) -> int:
        N = len(points)

        adj = collections.defaultdict(list) # adj: {key : [] }
        # node,all it's [cost,neighbours]

        for i in range(N):
            x1,y1 = points[i]
            for j in range(i+1, N):
                x2,y2 = points[j]
                dist = abs(x2-x1) + abs(y2-y1)
                adj[i].append([dist, j])
                adj[j].append([dist, i])
        # print(adj)
        
        result = 0
        visit = set()
        minHeap = [[0,0]] # [cost, point]

        while len(visit) < N:
            cost,i = heapq.heappop(minHeap)
            if i in visit:
                continue
            result += cost
            visit.add(i)
            for [neiCost, nei] in adj[i]:
                if nei not in visit:
                    heapq.heappush(minHeap, [neiCost, nei])
        return result

File: java/Q1584_Min_Cost_to_Connect_All_Points.java
class PointNode{
    public int cost;
    public int index;

    public PointNode(int cost,int index){
        this.cost = cost;
        this.index = index;
    }

    public String toString(){
        return "cost: " + cost + " index: " + index;
    }
}

class Solution {
    public int minCostConnectPoints(int[][] points) {
        if(points == null) return 0;
        
        PriorityQueue<PointNode> pq = new PriorityQueue(new Comparator<PointNode>(){
            @Override
            public int compare(PointNode p1, PointNode p2){
                return p1.cost-p2.cost;
            }
        });

        int N = points.length;
        Map<Integer, List<PointNode>> adj = new HashMap<>();

        for (int i = 0 ; i < N ; i++ ){
            for(int j = i+1 ; j < N ;j++){
                int cost = Math.abs(points[i][0]-points[j][0]) + Math.abs(points[i][1]-points[j][1]);

                List<PointNode> list_i = adj.getOrDefault( i, new LinkedList<PointNode>());
                list_i.add(new PointNode(cost,j));
                adj.put( i , list_i );

                List<PointNode> list_j = adj.getOrDefault( j, new LinkedList<PointNode>());
                list_j.add(new PointNode(cost,i));
                adj.put( j , list_j );
            }
        }

        pq.add(new PointNode(0,0));
        int res = 0;
        Set<Integer> visited = new HashSet<>();

        while(visited.size() < N){
            PointNode cur = pq.poll();
            if(visited.contains(cur.index)) continue;
            res += cur.cost;
            visited.add(cur.index);
            if(adj.get(cur.index)!=null){
                for( PointNode neib : adj.get(cur.index)){
                    if(visited.contains(neib.index)) continue;
                    pq.add(neib);
                }
            }
            
        }

        return res;
    }

}